packages/workflows/src/handlers/product/upsert-variant-prices.ts

```diff
@@ -0,0 +1,151 @@
+import { PricingTypes } from "@medusajs/types"
+
+import { WorkflowArguments } from "../../helper"
+
+type VariantPrice = {
+  id?: string
+  region_id?: string
+  currency_code: string
+  amount: number
+  min_quantity?: number
+  max_quantity?: number
+  rules: Record<string, string>
+}
+
+type HandlerInput = {
+  variantPricesMap: Map<string, VariantPrice[]>
+}
+
+export async function upsertVariantPrices({
+  container,
+  context,
+  data,
+}: WorkflowArguments<HandlerInput>) {
+  const { variantPricesMap } = data
+
+  const pricingModuleService = container.resolve("pricingModuleService")
+  const regionService = container.resolve("regionService")
+  const medusaApp = container.resolve("medusaApp")
+  const remoteQuery = container.resolve("remoteQuery")
+
+  const variables = {
+    variant_id: [...variantPricesMap.keys()],
+  }
+
+  const query = {
+    product_variant_price_set: {
+      __args: variables,
+      fields: ["variant_id", "price_set_id"],
+    },
+  }
+
+  const variantPriceSets = await remoteQuery(query)
+
+  const variantIdToPriceSetIdMap: Map<string, string> = new Map(
+    variantPriceSets.map((variantPriceSet) => [
+      variantPriceSet.variant_id,
+      variantPriceSet.price_set_id,
+    ])
+  )
+
+  const moneyAmountsToUpdate: PricingTypes.UpdateMoneyAmountDTO[] = []
+  const createdPriceSets: PricingTypes.PriceSetDTO[] = []
+  const ruleSetPricesToAdd: PricingTypes.CreatePricesDTO[] = []
+  const priceSetsToCreate: PricingTypes.CreatePriceSetDTO[] = []
+  const linksToCreate: any[] = []
+
+  for (const [variantId, prices = []] of variantPricesMap) {
+    const priceSetToCreate: PricingTypes.CreatePriceSetDTO = {
+      rules: [{ rule_attribute: "region_id" }],
+      prices: [],
+    }
+
+    for (const price of prices) {
+      if (price.id) {
+        moneyAmountsToUpdate.push({
+          id: price.id,
+          min_quantity: price.min_quantity,
+          max_quantity: price.max_quantity,
+          amount: price.amount,
+          currency_code: price.currency_code,
+        })
+      } else {
+        const variantPrice: PricingTypes.CreatePricesDTO = {
+          min_quantity: price.min_quantity,
+          max_quantity: price.max_quantity,
+          amount: price.amount,
+          currency_code: price.currency_code,
+          rules: {},
+        }
+
+        if (price.region_id) {
+          const region = await regionService.retrieve(price.region_id)
```

Comment:
suggestion: Save region id + currency in map to avoid retrieving region more than once

Reply:
Actually, you wouldn't include two prices with the same region id, so this shouldn't be an issue. I was considering this in a bulk-upsert scenario

Reply:
we could still consider listing all regions before entering the loop to still reduce the amount of queries, wdyt?

Reply:
yes, that's a great idea!